### ISSUE TITLE
fish gets stuck in history search mode if you press cmd+up-arrow

### ISSUE THREAD

**Issue by injust** [fish-shell/fish-shell#10584]
Using:
fish, version 3.7.1
macOS Sonoma 14.5
Ghostty terminal

When you press cmd+up-arrow, fish triggers history search mode as if you pressed up-arrow. But after you press down-arrow to exit history search mode, `commandline --search-mode` is still true for the current prompt.

I was unable to reproduce this in Terminal, kitty, or WezTerm. Might be because some terminals handle the cmd+up-arrow input themselves instead of sending it to the shell (e.g. Terminal uses it as a shortcut to jump to a mark).

I tested with the following `config.fish`:
```fish
# Modified from _atuin_bind_up at https://github.com/atuinsh/atuin/blob/95cc472037fcb3207b510e67f1a44af4e2a2cae9/crates/atuin/src/shell/atuin.fish#L55
function _bind_up
    # Fallback to fish's builtin up-or-search if we're in search or paging mode
    if commandline --search-mode
        echo search-mode
    else if commandline --paging-mode
        echo paging-mode
    else
        echo up-arrow
    end
end

if status is-interactive
    # How atuin binds the up-arrow, from `atuin init fish`
    bind -k up _bind_up
    bind \eOA _bind_up
    bind \e\[A _bind_up
    if bind -M insert > /dev/null 2>&1
        bind -M insert -k up _bind_up
        bind -M insert \eOA _bind_up
        bind -M insert \e\[A _bind_up
    end
end
```

https://github.com/fish-shell/fish-shell/assets/3387175/26cb921e-3a12-41b5-9144-869797db53c7

[asciicast.json](https://github.com/user-attachments/files/15942144/asciicast.json)

ref https://github.com/atuinsh/atuin/issues/2165

**Comment by mqudsi** [comment]
Just as an fyi, not too many of us have access to a (modern) macOS desktop for testing issues that arise with specific combinations of third party apps, so this might sit for a while if there's no repro on Linux.

**Comment by injust** [comment]
I don't have a Linux system to repro on, unfortunately.

I should mention that Ghostty is in closed beta right now (binaries available on macOS, source-only on Linux). I don't think this is a Ghostty issue, but if you need to test on Ghostty, I can ask to get you access.

**Comment by zanchey** [comment]
Can you try using <kbd>Up</kbd> and <kbd>Cmd</kbd>+<kbd>Up</kbd> in `fish_key_reader` to see what the difference is?

**Comment by injust** [comment]
`up`: `bind \e\[A 'do something'`
`cmd+up`: `bind \e\[1\;9A 'do something'`

**Comment by krobelus** [comment]
might be fixed on latest master; if not, can you share a Ghostty I can test with

**Comment by injust** [comment]
I built https://github.com/fish-shell/fish-shell/commit/4664a0b52f6ea92c8d19b935b52aafcdd3798905 and played around with it. I also noticed that `fish_key_reader` now prints `bind up 'do something'` for both <kbd>Up</kbd> and <kbd>cmd</kbd>+<kbd>Up</kbd>.

So I changed up my test `config.fish` to this:
```fish
function _print_mode
    if commandline --search-mode
        echo search-mode
    else if commandline --paging-mode
        echo paging-mode
    else
        echo up-arrow
    end
end

function _bind_up
    _print_mode
    up-or-search
end

if status is-interactive
    bind up _bind_up
end
```

It tells you if you're in search mode or not _before_ running `up-or-search`. For example, if you press <kbd>Up</kbd> twice, it prints `up-arrow` and then `search-mode`.

But if you press <kbd>Up</kbd> <kbd>Down</kbd> <kbd>Up</kbd>, the 2nd <kbd>Up</kbd> prints `search-mode` (i.e. <kbd>Down</kbd> didn't exit search mode!)

Basically: <kbd>cmd</kbd>+<kbd>Up</kbd> was a red herring (hah). It doesn't matter whether you _enter_ search mode by pressing <kbd>Up</kbd> or <kbd>cmd</kbd>+<kbd>Up</kbd>. The actual problem is that <kbd>Down</kbd> never _exits_ search mode.

And because you don't actually need <kbd>cmd</kbd>+<kbd>Up</kbd> to repro this, I repro'd it in Terminal.app, and it probably works in your preferred terminal emulator, too.

If this isn't clear, let me know and I can grab a screen recording later.

**Comment by krobelus** [comment]
> The actual problem is that Down never exits search mode.

Yeah. The accumulated search results will be cached until you move the cursor etc., that's why.

For context: the motivation is that `atuin` binds up arrow to

```fish
function _atuin_bind_up
    # Fallback to fish's builtin up-or-search if we're in search or paging mode
    if commandline --search-mode; or commandline --paging-mode
        up-or-search
        return
    end

    ...
end
```

the workaround is to move the cursor.

Possible solutions:
- Provide `commandline --search-is-at-end` which atuin can use
- Exit history search when at end (discarding results)

**Comment by injust** [comment]
Based on https://fishshell.com/docs/current/cmds/commandline.html:

> **-S** or **--search-mode**
> Evaluates to true if the commandline is performing a history search.

Exiting history search at the end _feels_ like the natural thing to do.

This sounds like an implementation decision now (caching upside of staying in search mode vs. one more command and clunkiness of the interface).

**Comment by krobelus** [comment]
To recap, the problem scenario is that you

0. use something like `bind up _atuin_bind_up`
1. press `ctrl-p` to start a native history search
2. press `ctrl-n` to move back to the original commandline
3. press `up` with the intent to start atuin history search

However we are still in search mode[*]; atuin detects that and delegates back to `up-or-search`.

Given that you still need an alternative binding like `ctrl-p` to be able to start a history search, it seems weird to overload `up` like that. 

Unless I missed a scenario where this actually helps, it seems me that `_atuin_bind_up` should simply start by terminating any native history search.

[*]: besides caching, there is a difference with multiline search strings.
     Not very important at all but I'd prefer not to add a hack if possible
